Fraud is one of the major issues we come up majorly in banks, life insurance, health insurance, and many others. These major frauds are dependent on the person who is trying to sell you the fake product or service, if you are matured enough to decide what is wrong then you will never get into any fraud transactions. But one such fraud that has been increasing a lot these days is fraud in making payments. This project aims to provide a solution with machine learing.

In [1]:
import os

In [2]:
#checking current directory path
os.getcwd()

'C:\\Users\\DELL\\Desktop\\py'

In [4]:
#importing pandas and numpy
import pandas as pd
import numpy as np

In [53]:
#reading the data file
df = pd.read_csv("payment_fraud.csv")

In [54]:
#seeing no.of features,target variabke and amount data present
df

accountAgeDays  numItems  localTime paymentMethod  \
0                  29         1   4.745402        paypal   
1                 725         1   4.742303   storecredit   
2                 845         1   4.921318    creditcard   
3                 503         1   4.886641    creditcard   
4                2000         1   5.040929    creditcard   
...               ...       ...        ...           ...   
39216             986         1   4.836982    creditcard   
39217            1647         1   4.876771    creditcard   
39218            1591         1   4.742303    creditcard   
39219             237         1   4.921318    creditcard   
39220             272         1   5.040929        paypal   

       paymentMethodAgeDays  label  
0                 28.204861      0  
1                  0.000000      0  
2                  0.000000      0  
3                  0.000000      0  
4                  0.000000      0  
...                     ...    ...  
39216              0.000000      0  
39217            377.930556      0  
39218              0.000000      0  
39219            236.082639      0  
39220              0.000694      0  

[39221 rows x 6 columns]

In [18]:
#columns names
df.columns

Index(['accountAgeDays', 'numItems', 'localTime', 'paymentMethod',
       'paymentMethodAgeDays', 'label'],
      dtype='object')

In [19]:
#size of the data(number of rows and columns)
df.shape

(39221, 6)

In [15]:
#cheking if data contains any missing value or not
df.isnull().sum()

accountAgeDays          0
numItems                0
localTime               0
paymentMethod           0
paymentMethodAgeDays    0
label                   0
dtype: int64

#### Since no missing values are there in the dataset, we will proceed to model building now. Since this is a problem of binary classification, I have used Logistic regression.

In [23]:
#importing packages from sklearn for  train-test split and model building
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score, confusion_matrix

In [56]:
#splitting the data into test and train set
X_train, X_test, y_train, y_test = train_test_split(df.drop(["label","paymentMethod"],axis=1),df["label"], test_size=0.30, random_state=16)

In [57]:
#creating instance for logistic regressor
lr = LogisticRegression()

In [58]:
#fitting the LR model
data = lr.fit(X_train, y_train)

In [59]:
#making predictions on the test set
y_pred = data.predict(X_test)

In [62]:
#checking accuracy score for the trained model
print(accuracy_score(y_test, y_pred))

1.0


In [63]:
# evaluationg the fraud detection model using confusion matrix
confusion_matrix(y_test,y_pred)

array([[11603,     0],
       [    0,   164]], dtype=int64)

#### Out of the total transactions in the dataset, 164  are correctly classified as fraudulent  and 11603 no fraudulent transactions

In [69]:
y_test.loc[df.label==1]

34619    1
17492    1
23803    1
884      1
6819     1
        ..
38344    1
10638    1
30475    1
22318    1
37769    1
Name: label, Length: 164, dtype: int64

#### Note:- This model can further be experimented using payment Method feature to see how much it helps in predicting the fraudulent transactions. Since this is a catogorical feature, we have to appy encoding tecnniques like one hot encoding since number of classes is 3 for the payment method columns.

In [76]:
#number of classes present for the payment method column
df.paymentMethod.unique()

array(['paypal', 'storecredit', 'creditcard'], dtype=object)